psl-core/src/main/java/org/linqs/psl/util/ModelDataCollector.java

```diff
@@ -0,0 +1,332 @@
+        @Override
+        public void run() {
+            outputJSON();
+        }
+    }
+
+    public static class ModelData {
+        public Map<String, Float> truthMap;
+        public Map<String, Map<String, Object>> rules;
+        public Map<String, Map<String, Object>> groundRules;
+        public Map<String, Map<String, Object>> groundAtoms;
+
+        public ModelData() {
+            truthMap = new HashMap<String, Float>();
+            rules = new HashMap<String, Map<String, Object>>();
+            groundRules = new HashMap<String, Map<String, Object>>();
+            groundAtoms = new HashMap<String, Map<String, Object>>();
+        }
+    }
+
+    public static void setOutputPath(String path) {
+        outputPath = path;
+    }
+
+    // Takes in a prediction truth pair and adds it to the Truth Map.
+    public static void addTruth(GroundAtom target, float truthVal ) {
+        String groundAtomID = Integer.toString(System.identityHashCode(target));
+        modelData.truthMap.put(groundAtomID, truthVal);
+    }
+
+    // Adds a ground rules dissatisfaction / infeasibility.
+    public static void dissatisfactionPerGroundRule(GroundRuleStore groundRuleStore) {
+        for (GroundRule groundRule : groundRuleStore.getGroundRules()) {
+            String strGroundRuleId = Integer.toString(System.identityHashCode(groundRule));
+            Map<String, Object> groundRuleObj = modelData.groundRules.get(strGroundRuleId);
+            if (groundRule instanceof WeightedGroundRule) {
+                WeightedGroundRule weightedGroundRule = (WeightedGroundRule) groundRule;
+                groundRuleObj.put("dissatisfaction", weightedGroundRule.getIncompatibility());
+            } else {
+                UnweightedGroundRule unweightedGroundRule = (UnweightedGroundRule) groundRule;
+                groundRuleObj.put("dissatisfaction", unweightedGroundRule.getInfeasibility());
+            }
+        }
+    }
+
+    // Decorates an entry in a Formula to have constants or negations marked.
+    public static Object decorateFormula(String groundAtom, GroundRule groundRule, boolean negation){
+        for (GroundAtom atom : groundRule.getAtoms()) {
+            if (groundAtom.contains(atom.toString())){
+                if (negation) {
+                    Integer[] groundAtomObj = {System.identityHashCode(atom), 1};
+                    return groundAtomObj;
+                } else {
+                    return System.identityHashCode(atom);
+                }
+            }
+        }
+        return null;
+    }
+
+    // Parses through an atom object, and fills out variables with their proper values.
+    public static String parseAtom (Formula f, Map<String, String> varConstMap) {
+        Atom atom = (Atom) f;
+        String groundedAtom = atom.toString();
+        Term[] arguments = atom.getArguments();
+        for (Term t : arguments){
+            if (t instanceof Variable) {
```

Comment:
Invert this conditional.

The more indented the code is, the more logical blocks deep you are (on this line, it is: `class -> method -> loop -> conditional`).
For every logical block, a programmer has to remember the context that put them into that block.
When possible, you want to reduce the burden on the reader by staying shallow.
In this case, you can do that by instead seeing if `t` is **not** a variable and then continuing.
Then the rest of the loop doesn't need to be inside that conditional.
```
if (!(t instanceof Variable)) {
    continue;
}
...
```